sql: improve routine dependency tracking

DrewKimball opened the pull request:
#### sql: track routine dependencies on columns in RETURNING clause

Previously, while building a routine, we did not explicitly add dependencies
to columns referenced in the RETURNING clause of mutations within the body
statements. This was not a problem for INSERT or UPSERT due to #145098, but
could cause a routine with UPDATE or DELETE statements to be broken after
a referenced column was dropped. This commit fixes the bug by tracking
column references in the RETURNING clause.

Fixes #146414

Release note (bug fix): Fixed a bug that allowed a column to be dropped
from a table even if it was referenced in the RETURNING clause of an UPDATE
or DELETE statement in a routine. In releases prior to v25.3, the fix will
be off by default, and can be enabled by setting the session variable
`use_improved_routine_dependency_tracking`.

#### opt: do not add unnecessary columns to routine deps

Previously, computed columns would be added to the depended-on columns
list for a routine that performs an INSERT. This is unnecessary, since
values for computed columns are automatically generated when the routine
is invoked depending on the table schema at the time of invocation.
Because of this behavior, a routine with an INSERT statement previously
prevented dropping computed columns from the target table, including the
implicit column used for hash-sharded indexes.

This commit skips computed columns when adding insert target columns to
the routine's dependency list. Note that there was already logic to do
this for columns with default values. This commit also adds a session
variable `use_improved_routine_dependency_tracking` to control the new
behavior, default off for backports. The follow-up will flip the variable
on for master.

Fixes #145098

Release note (sql change): Fixed a bug that caused a routine with an
INSERT statement to unnecessarily block dropping a hash-sharded index
or computed column on the target table. Note that the fix applies only
to newly-created routines. In releases prior to v25.3, the fix will
be off by default, and can be enabled by setting the session variable
`use_improved_routine_dependency_tracking`.

#### sql: add session setting for routine dependency improvements

This commit adds a session var `use_improved_routine_dependency_tracking`
which enables the improvements in the previous two commits (do not add
dependencies on unnecessary columns, and track dependencies in the
RETURNING clause). The var is off by default in this commit for backports.

Informs #145098
Informs #146414

Release note: None

#### opt: enable improved routine dependency tracking

This commit flips `use_improved_routine_dependency_tracking` to on by
default. This means that newly-created routines will not add unnecessary
implicit insert columns to the list of dependencies, and will correctly
track dependencies in the RETURNING clause of UPDATE and DELETE statements.

Informs #145098
Informs #146414

Release note: None

DrewKimball committed:
opt: enable improved routine dependency tracking

This commit flips `use_improved_routine_dependency_tracking` to on by
default. This means that newly-created routines will not add unnecessary
implicit insert columns to the list of dependencies, and will correctly
track dependencies in the RETURNING clause of UPDATE and DELETE statements.

Informs #145098
Informs #146414

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/146250)
<!-- Reviewable:end -->

yuzefovich reviewed (approved):
Nice find! <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 10 of 10 files at r1, 4 of 4 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/146250#-:-OPc8oe-7cNef9v-ldwX:boyyrm)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @rytaft)

___
*[`pkg/sql/logictest/testdata/logic_test/udf_insert` line 293 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/146250#-OPc8DhY-qgFzjUBrPMk:-OPc8DhY-qgFzjUBrPMl:b-40hzgj) ([raw file](https://github.com/cockroachdb/cockroach/blob/2f1f4f8f20ddf2c3127b9164322f2989cea4f6bc/pkg/sql/logictest/testdata/logic_test/udf_insert#L293)):*
> ```
> 
> statement ok
> SELECT f145098();
> ```

nit: perhaps we should verify the contents of `t_computed` after each UDF invocation (at least the number of rows in the table)?


<!-- Sent from Reviewable.io -->

michae2 reviewed (approved):
Suggested a few more test ideas, but <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

Thanks for fixing this!

> Reviewed 10 of 10 files at r1, 4 of 4 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/146250#-:-OPgrpUW9ovA41a9iTEH:b-vc9eb5)* status: <span>:shipit:</span> complete! 2 of 0 LGTMs obtained (waiting on @DrewKimball and @rytaft)

___
*[`pkg/sql/logictest/testdata/logic_test/udf_delete` line 322 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/146250#-OPgzGgb7P1AI43wNpz0:-OPgzGgc03fzFXQjVkXG:ba4rk69) ([raw file](https://github.com/cockroachdb/cockroach/blob/2f1f4f8f20ddf2c3127b9164322f2989cea4f6bc/pkg/sql/logictest/testdata/logic_test/udf_delete#L322)):*
> ```
> statement ok
> CREATE FUNCTION f_delete() RETURNS INT LANGUAGE SQL AS $$
>   DELETE FROM table_drop WHERE a = b - 1;
> ```

suggestion: Might be good to also test that `RETURNING` a hidden hash column _does_ add it to the dependency set. (One idea is to add a second hash index that cannot be dropped for this reason.)

Also might be good to test that a predicate on a hidden hash column adds it to the dependency set.
___
*[`pkg/sql/logictest/testdata/logic_test/udf_insert` line 279 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/146250#-OPgy0EB9GhrzUrRfOJk:-OPgy0EB9GhrzUrRfOJl:b-ayqvf3) ([raw file](https://github.com/cockroachdb/cockroach/blob/2f1f4f8f20ddf2c3127b9164322f2989cea4f6bc/pkg/sql/logictest/testdata/logic_test/udf_insert#L279)):*
> ```
> 
> # An insert routine created with use_improved_routine_dependency_tracking = true
> # should not depend on columns other than the target column "a". 
> ```

typo: trailing space


<!-- Sent from Reviewable.io -->

DrewKimball reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/146250)* status: <span>:shipit:</span> complete! 2 of 0 LGTMs obtained (waiting on @michae2 and @rytaft)

___
*[`pkg/sql/logictest/testdata/logic_test/udf_delete` line 322 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/146250#-OPgzGgb7P1AI43wNpz0:-OPmNl6m8W48Zt9UwNrh:b2r9ac0) ([raw file](https://github.com/cockroachdb/cockroach/blob/2f1f4f8f20ddf2c3127b9164322f2989cea4f6bc/pkg/sql/logictest/testdata/logic_test/udf_delete#L322)):*
>Might be good to also test that `RETURNING` a hidden hash column _does_ add it to the dependency set.

That's a great idea. It uncovered a bug in my fix, and also [an existing one](https://github.com/cockroachdb/cockroach/issues/146414) in update and delete. I'll work on another commit to fix that.


<!-- Sent from Reviewable.io -->

michae2 commented:
Just checking, is this ready for another look?

DrewKimball committed:
sql: add session setting for routine dependency improvements

This commit adds a session var `use_improved_routine_dependency_tracking`
which enables the improvements in the previous two commits (do not add
dependencies on unnecessary columns, and track dependencies in the
RETURNING clause). The var is off by default in this commit for backports.

Informs #145098
Informs #146414

Release note: None

DrewKimball committed:
sql: track routine dependencies on columns in RETURNING clause

Previously, while building a routine, we did not explicitly add dependencies
to columns referenced in the RETURNING clause of mutations within the body
statements. This was not a problem for INSERT or UPSERT due to #145098, but
could cause a routine with UPDATE or DELETE statements to be broken after
a referenced column was dropped. This commit fixes the bug by tracking
column references in the RETURNING clause.

Fixes #146414

Release note (bug fix): Fixed a bug that allowed a column to be dropped
from a table even if it was referenced in the RETURNING clause of an UPDATE
or DELETE statement in a routine. In releases prior to v25.3, the fix will
be off by default, and can be enabled by setting the session variable
`use_improved_routine_dependency_tracking`.

DrewKimball committed:
opt: do not add unnecessary columns to routine deps

Previously, computed columns would be added to the depended-on columns
list for a routine that performs an INSERT. This is unnecessary, since
values for computed columns are automatically generated when the routine
is invoked depending on the table schema at the time of invocation.
Because of this behavior, a routine with an INSERT statement previously
prevented dropping computed columns from the target table, including the
implicit column used for hash-sharded indexes.

This commit skips computed columns when adding insert target columns to
the routine's dependency list. Note that there was already logic to do
this for columns with default values. This commit also adds a session
variable `use_improved_routine_dependency_tracking` to control the new
behavior, default off for backports. The follow-up will flip the variable
on for master.

Fixes #145098

Release note (sql change): Fixed a bug that caused a routine with an
INSERT statement to unnecessarily block dropping a hash-sharded index
or computed column on the target table. Note that the fix applies only
to newly-created routines. In releases prior to v25.3, the fix will
be off by default, and can be enabled by setting the session variable
`use_improved_routine_dependency_tracking`.

DrewKimball commented:
Sorry for the delay, this is RFAL now. The first commit fixes https://github.com/cockroachdb/cockroach/issues/145098, the second fixes https://github.com/cockroachdb/cockroach/issues/146414, and the last two add a session setting that toggles the new behavior. We'll backport with the setting off by default (which is why I removed the backport label).

yuzefovich reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 3 of 15 files at r3, 4 of 12 files at r5, 8 of 8 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/146250#-:-OS6MGxc2-hu1QPsc9Ix:bnfp4nl)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @DrewKimball, @michae2, and @rytaft)

___
*[`pkg/sql/logictest/testdata/logic_test/udf_insert` line 278 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/146250#-OS6Idao6tuTvgzu3XDT:-OS6Idao6tuTvgzu3XDU:b2yep0y) ([raw file](https://github.com/cockroachdb/cockroach/blob/b4fa4829072a3c7a3f26a98b46e492b96995712b/pkg/sql/logictest/testdata/logic_test/udf_insert#L278)):*
> ```
> statement ok
> CREATE FUNCTION f146414() RETURNS INT LANGUAGE SQL AS $$
>   UPSERT INTO t146414 (a) VALUES (100) RETURNING b;
> ```

nit: should this one be INSERT?
___
*[`pkg/sql/opt/optbuilder/util.go` line 868 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/146250#-OS6JWcL0HklVnricS6j:-OS6JWcL0HklVnricS6k:b-b2rs2e) ([raw file](https://github.com/cockroachdb/cockroach/blob/b4fa4829072a3c7a3f26a98b46e492b96995712b/pkg/sql/opt/optbuilder/util.go#L868)):*
> ```Go
> 	trackSchemaDepsForReturning := b.trackSchemaDeps &&
> 		b.evalCtx.SessionData().UseImprovedRoutineDependencyTracking
> 	if trackSchemaDepsForReturning {
> ```

nit: personal preference, but I'd find the code easier to follow if we split up the responsibility of populating `cols` from conditionally building `dep`. In other words,
```
for ... {
  // build s.cols
}
if trackSchemaDepsForReturning {
  for ... {
    // build dep
  }
  b.schemaDeps = append(...)
}
```
seems cleaner to me, and the cost of an extra loop iteration should be negligible.


<!-- Sent from Reviewable.io -->

DrewKimball reviewed (commented):
TFYRs!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/146250#-:-OSPsXiWCV1w9_N7-b14:bz5phge)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @michae2, @rytaft, and @yuzefovich)

___
*[`pkg/sql/opt/optbuilder/util.go` line 868 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/146250#-OS6JWcL0HklVnricS6j:-OSPs11B7OJiX-mycC2w:bcd6uar) ([raw file](https://github.com/cockroachdb/cockroach/blob/b4fa4829072a3c7a3f26a98b46e492b96995712b/pkg/sql/opt/optbuilder/util.go#L868)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: personal preference, but I'd find the code easier to follow if we split up the responsibility of populating `cols` from conditionally building `dep`. In other words,
```
for ... {
  // build s.cols
}
if trackSchemaDepsForReturning {
  for ... {
    // build dep
  }
  b.schemaDeps = append(...)
}
```
seems cleaner to me, and the cost of an extra loop iteration should be negligible.
</blockquote></details>

Nice idea, Done.
___
*[`pkg/sql/logictest/testdata/logic_test/udf_insert` line 279 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/146250#-OPgy0EB9GhrzUrRfOJk:-OSPsWgX9NXxeEkf61hk:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/2f1f4f8f20ddf2c3127b9164322f2989cea4f6bc/pkg/sql/logictest/testdata/logic_test/udf_insert#L279)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

typo: trailing space
</blockquote></details>

Done.
___
*[`pkg/sql/logictest/testdata/logic_test/udf_insert` line 293 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/146250#-OPc8DhY-qgFzjUBrPMk:-OPlP6AO4fjzRCfjjLHc:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/2f1f4f8f20ddf2c3127b9164322f2989cea4f6bc/pkg/sql/logictest/testdata/logic_test/udf_insert#L293)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: perhaps we should verify the contents of `t_computed` after each UDF invocation (at least the number of rows in the table)?
</blockquote></details>

Done.
___
*[`pkg/sql/logictest/testdata/logic_test/udf_insert` line 278 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/146250#-OS6Idao6tuTvgzu3XDT:-OSPpwoc6igqC9D_-kAN:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b4fa4829072a3c7a3f26a98b46e492b96995712b/pkg/sql/logictest/testdata/logic_test/udf_insert#L278)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: should this one be INSERT?
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

michae2 reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Thanks for doing this!

> Reviewed 3 of 15 files at r3, 12 of 12 files at r7, 3 of 8 files at r10, 10 of 10 files at r11, 13 of 13 files at r12, 9 of 9 files at r13, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/146250#-:-OSQRdW4DWSt5bKYE20f:b-pt8m11)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @rytaft and @yuzefovich)

___
*[`pkg/sql/vars.go` line 4039 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/146250#-OSQeYdQ7mMa-NBgUs_8:-OSQeYdQ7mMa-NBgUs_9:bsz4fej) ([raw file](https://github.com/cockroachdb/cockroach/blob/c99cb7448cdb9b8b5f0f8d5cdf72a97bf1261053/pkg/sql/vars.go#L4039)):*
> ```Go
> 
> 	// CockroachDB extension.
> 	`use_improved_routine_dependency_tracking`: {
> ```

Does this need to be added to the memo staleness check?


<!-- Sent from Reviewable.io -->

DrewKimball reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/146250)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @michae2, @rytaft, and @yuzefovich)

___
*[`pkg/sql/vars.go` line 4039 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/146250#-OSQeYdQ7mMa-NBgUs_8:-OSQfWZ84UKvTRZj4Zw7:b-beys0b) ([raw file](https://github.com/cockroachdb/cockroach/blob/c99cb7448cdb9b8b5f0f8d5cdf72a97bf1261053/pkg/sql/vars.go#L4039)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

Does this need to be added to the memo staleness check?
</blockquote></details>

This applies to routine creation, for which we never cache plans. So that won't be necessary.


<!-- Sent from Reviewable.io -->

DrewKimball commented:
TFTRs!

bors r=yuzefovich,michae2
